### ISSUE TITLE
Cannot install update, in VSCode Insiders

### ISSUE THREAD

**Issue by jtsom** [OmniSharp/omnisharp-vscode#4922]
## Issue Description ##

Error while trying to install latest update.

## Steps to Reproduce ##
Try to install latest version.

## Expected Behavior ##
Version installs successfully.

## Actual Behavior ##
Installation fails:

```
[2021-12-02 10:09:55.790] [renderer2] [error] [Extension Host] (node:10324) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(Use `Code - Insiders --trace-deprecation ...` to show where the warning was created)
[2021-12-02 10:10:28.806] [renderer2] [error] Problem invoking 'GetCodeActions' on OmniSharp server: Error: Request /v2/getcodeactions cancelled, id: 10
[2021-12-02 10:10:29.805] [renderer2] [error] Problem invoking 'GetCodeActions' on OmniSharp server: Error: Request /v2/getcodeactions cancelled, id: 18
[2021-12-02 10:10:30.790] [renderer2] [error] Problem invoking 'GetCodeActions' on OmniSharp server: Error: Request /v2/getcodeactions cancelled, id: 21
[2021-12-02 10:10:31.829] [renderer2] [error] Problem invoking 'GetCodeActions' on OmniSharp server: Error: Request /v2/getcodeactions cancelled, id: 23
[2021-12-03 08:57:40.253] [renderer2] [error] Cannot install 'ms-dotnettools.csharp' extension because of version mismatch in Marketplace: Invalid: Cannot install 'ms-dotnettools.csharp' extension because of version mismatch in Marketplace
    at A.checkAndGetCompatibleVersion (vscode-file://vscode-app/c:/Users/jtsombakos/AppData/Local/Programs/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:13083)
    at async A.installFromGallery (vscode-file://vscode-app/c:/Users/jtsombakos/AppData/Local/Programs/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:2928)
[2021-12-03 08:57:42.155] [renderer2] [error] ["Invalid: Cannot install 'ms-dotnettools.csharp' extension because of version mismatch in Marketplace","    at A.checkAndGetCompatibleVersion (vscode-file://vscode-app/c:/Users/jtsombakos/AppData/Local/Programs/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:13083)","    at async A.installFromGallery (vscode-file://vscode-app/c:/Users/jtsombakos/AppData/Local/Programs/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:2928)"]
[2021-12-03 08:57:42.157] [renderer2] [error] Cannot install 'ms-dotnettools.csharp' extension because of version mismatch in Marketplace: Invalid: Cannot install 'ms-dotnettools.csharp' extension because of version mismatch in Marketplace
    at A.checkAndGetCompatibleVersion (vscode-file://vscode-app/c:/Users/jtsombakos/AppData/Local/Programs/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:13083)
    at async A.installFromGallery (vscode-file://vscode-app/c:/Users/jtsombakos/AppData/Local/Programs/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:2928)
[2021-12-03 08:57:47.003] [renderer2] [error] ["Invalid: Cannot install 'ms-dotnettools.csharp' extension because of version mismatch in Marketplace","    at A.checkAndGetCompatibleVersion (vscode-file://vscode-app/c:/Users/jtsombakos/AppData/Local/Programs/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:13083)","    at async A.installFromGallery (vscode-file://vscode-app/c:/Users/jtsombakos/AppData/Local/Programs/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:2928)"]
[2021-12-03 08:57:47.005] [renderer2] [error] Cannot install 'ms-dotnettools.csharp' extension because of version mismatch in Marketplace: Invalid: Cannot install 'ms-dotnettools.csharp' extension because of version mismatch in Marketplace
    at A.checkAndGetCompatibleVersion (vscode-file://vscode-app/c:/Users/jtsombakos/AppData/Local/Programs/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:13083)
    at async A.installFromGallery (vscode-file://vscode-app/c:/Users/jtsombakos/AppData/Local/Programs/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:2928)
[2021-12-03 08:57:47.879] [renderer2] [error] ["Invalid: Cannot install 'ms-dotnettools.csharp' extension because of version mismatch in Marketplace","    at A.checkAndGetCompatibleVersion (vscode-file://vscode-app/c:/Users/jtsombakos/AppData/Local/Programs/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:13083)","    at async A.installFromGallery (vscode-file://vscode-app/c:/Users/jtsombakos/AppData/Local/Programs/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:2928)"]
[2021-12-03 08:57:48.560] [renderer2] [error] ["Invalid: Cannot install 'ms-dotnettools.csharp' extension because of version mismatch in Marketplace","    at A.checkAndGetCompatibleVersion (vscode-file://vscode-app/c:/Users/jtsombakos/AppData/Local/Programs/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:13083)","    at async A.installFromGallery (vscode-file://vscode-app/c:/Users/jtsombakos/AppData/Local/Programs/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:2928)"]
[2021-12-03 08:57:48.579] [renderer2] [error] Cannot install 'ms-dotnettools.csharp' extension because of version mismatch in Marketplace: Invalid: Cannot install 'ms-dotnettools.csharp' extension because of version mismatch in Marketplace
    at A.checkAndGetCompatibleVersion (vscode-file://vscode-app/c:/Users/jtsombakos/AppData/Local/Programs/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:13083)
    at async A.installFromGallery (vscode-file://vscode-app/c:/Users/jtsombakos/AppData/Local/Programs/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:2928)
[2021-12-03 08:57:49.283] [renderer2] [error] ["Invalid: Cannot install 'ms-dotnettools.csharp' extension because of version mismatch in Marketplace","    at A.checkAndGetCompatibleVersion (vscode-file://vscode-app/c:/Users/jtsombakos/AppData/Local/Programs/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:13083)","    at async A.installFromGallery (vscode-file://vscode-app/c:/Users/jtsombakos/AppData/Local/Programs/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:2928)"]

```
### OmniSharp log ###
<details>Post the output from Output-->OmniSharp log here</details>

### C# log ###
<details>Post the output from Output-->C# here</details>

## Environment information ##

**VSCode version**: 1.63.0-insider
**C# Extension**: 1.23.16

<details><summary>Dotnet Information</summary>
.NET SDK (reflecting any global.json):
 Version:   6.0.100
 Commit:    9e8b04bbff

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19044
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.100\

Host (useful for support):
  Version: 6.0.0
  Commit:  4822e3c3aa

.NET SDKs installed:
  5.0.403 [C:\Program Files\dotnet\sdk]
  6.0.100 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.21 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.12 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.21 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.21 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.12 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download
</details>
<details><summary>Visual Studio Code Extensions</summary>

|Extension|Author|Version|
|---|---|---|
|azure-account|ms-vscode|0.9.11|
|csharp|ms-dotnettools|1.23.16|
|git-graph|mhutchie|1.30.0|
|github-vscode-theme|GitHub|5.1.0|
|hexeditor|ms-vscode|1.8.2|
|hl7tools|RobHolme|1.13.2|
|horizon-extended|lancewilhelm|0.1.5|
|material-icon-theme|PKief|4.11.0|
|msbuild-project-tools|tintoy|0.4.1|
|night-owl|sdras|2.0.1|
|quicktype|quicktype|12.0.46|
|TabOut|albert|0.2.1|
|vsc-community-material-theme|Equinusocio|1.4.4|
|vscode-azurefunctions|ms-azuretools|1.6.0|
|vscode-azureresourcegroups|ms-azuretools|0.4.0|
|vscode-azurestorage|ms-azuretools|0.12.1|
|vscode-base64|adamhartford|0.1.0|
|vscode-thunder-client|rangav|1.9.1|
|vscodeintellicode|VisualStudioExptTeam|1.2.14|;

</details>

**Comment by phillipo6** [comment]
Got the same problem!

**Comment by jtsom** [comment]
Completely removing the extension, including deleting it from the .vscode-insiders/extensions directory, then manually downloading and installing the .vsix, seems to only install v1.23.16. Claims to want to update again, and again fails.

**Comment by jtsom** [comment]
After downloading "csharp-1.23.17-win32-x64"  from the repo (vs. the file that was offered to me in the error popup "ms-dotnettools.csharp-1.23.17.vsix"), it seemed to load... but the output log still shows an error that it cannot load:
```
[2021-12-03 09:17:13.160] [renderer1] [error] [Extension Host] (node:20008) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
(Use `Code - Insiders --trace-deprecation ...` to show where the warning was created)
[2021-12-03 09:17:16.232] [renderer1] [error] Cannot install 'ms-dotnettools.csharp' extension because of version mismatch in Marketplace: Invalid: Cannot install 'ms-dotnettools.csharp' extension because of version mismatch in Marketplace
    at A.checkAndGetCompatibleVersion (vscode-file://vscode-app/c:/Users/jtsombakos/AppData/Local/Programs/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:13083)
    at async A.installFromGallery (vscode-file://vscode-app/c:/Users/jtsombakos/AppData/Local/Programs/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:2928)
[2021-12-03 09:17:24.314] [renderer1] [error] ["Invalid: Cannot install 'ms-dotnettools.csharp' extension because of version mismatch in Marketplace","    at A.checkAndGetCompatibleVersion (vscode-file://vscode-app/c:/Users/jtsombakos/AppData/Local/Programs/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:13083)","    at async A.installFromGallery (vscode-file://vscode-app/c:/Users/jtsombakos/AppData/Local/Programs/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:2928)"]
[2021-12-03 09:17:24.316] [renderer1] [error] Cannot install 'ms-dotnettools.csharp' extension because of version mismatch in Marketplace: Invalid: Cannot install 'ms-dotnettools.csharp' extension because of version mismatch in Marketplace
    at A.checkAndGetCompatibleVersion (vscode-file://vscode-app/c:/Users/jtsombakos/AppData/Local/Programs/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:13083)
    at async A.installFromGallery (vscode-file://vscode-app/c:/Users/jtsombakos/AppData/Local/Programs/Microsoft%20VS%20Code%20Insiders/resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:2928)
```

**Comment by MariaSolOs** [comment]
I have the same issue ("version mismatch in the marketplace"):
```
[2021-12-03 16:01:33.892] [renderer1] [error] ["Invalid: Cannot install 'ms-dotnettools.csharp' extension because of version mismatch in Marketplace","    at A.checkAndGetCompatibleVersion (vscode-file://vscode-app/Applications/Visual%20Studio%20Code%20-%20Insiders.app/Contents/Resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:13083)","    at async A.installFromGallery (vscode-file://vscode-app/Applications/Visual%20Studio%20Code%20-%20Insiders.app/Contents/Resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:2928)"]
[2021-12-03 16:01:33.930] [renderer1] [error] Cannot install 'ms-dotnettools.csharp' extension because of version mismatch in Marketplace: Invalid: Cannot install 'ms-dotnettools.csharp' extension because of version mismatch in Marketplace
    at A.checkAndGetCompatibleVersion (vscode-file://vscode-app/Applications/Visual%20Studio%20Code%20-%20Insiders.app/Contents/Resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:13083)
    at async A.installFromGallery (vscode-file://vscode-app/Applications/Visual%20Studio%20Code%20-%20Insiders.app/Contents/Resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:2928)
```

**Comment by JMackCo** [comment]
Same issue here

**Comment by JMackCo** [comment]
Found vscode still running as service. I did an end task on vscode and then was able to install the C# extension. However before that I did clean out the Contents/Resources/app/out

**Comment by HeckTwrcha** [comment]
The upgrade from v1.23.16 to v1.23.17 works fine for me for VS Code however it failed for VS Code-Insider with the version mismatch error. Then when I try to install the extension manually "ms-dotnettools.csharp-1.23.17.vsix" downloaded from the marketplace it installed v.1.23.16 instead.

**Comment by FooqX** [comment]
Same issue. Please fix this.

**Comment by yamachu** [comment]
I rewrote part of `checkAndGetCompatibleVersion` in VSCode Insiders with following.

```js
throw new d.ExtensionManagementError(`Cannot install '${a.identifier.id}' extension because of version mismatch in Marketplace, ${u.version}: ${o.version}
```

`u.version` means `manifest.version`, and `o.version` means `compatibleExtension.version`.

https://github.com/microsoft/vscode/blob/bb0045c54f431faee661a4d5590a044097938e63/src/vs/platform/extensionManagement/common/abstractExtensionManagementService.ts#L423

```
[2021-12-06 11:39:34.049] [renderer1] [error] ["Invalid: Cannot install 'ms-dotnettools.csharp' extension because of version mismatch in Marketplace, 1.23.16: 1.23.17","    at A.checkAndGetCompatibleVersion (vscode-file://vscode-app/Applications/Visual%20Studio%20Code%20-%20Insiders.app/Contents/Resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:13083)","    at async A.installFromGallery (vscode-file://vscode-app/Applications/Visual%20Studio%20Code%20-%20Insiders.app/Contents/Resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:2928)"]
[2021-12-06 11:39:34.050] [renderer1] [error] Cannot install 'ms-dotnettools.csharp' extension because of version mismatch in Marketplace, 1.23.16: 1.23.17: Invalid: Cannot install 'ms-dotnettools.csharp' extension because of version mismatch in Marketplace, 1.23.16: 1.23.17
    at A.checkAndGetCompatibleVersion (vscode-file://vscode-app/Applications/Visual%20Studio%20Code%20-%20Insiders.app/Contents/Resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:13083)
    at async A.installFromGallery (vscode-file://vscode-app/Applications/Visual%20Studio%20Code%20-%20Insiders.app/Contents/Resources/app/out/vs/code/electron-browser/sharedProcess/sharedProcessMain.js:48:2928)
```

And I tried `Try Downloading Manually...` .
<img width="463" alt="スクリーンショット 2021-12-06 12 01 38" src="https://user-images.githubusercontent.com/1955233/144780307-ad0e9320-1e9c-4336-91f6-562867e9c366.png">

1. unzip `ms-dotnettools.csharp-1.23.17.vsix`
2. grep 1.23.16 extension.vsixmanifest

```
    <Identity Language="en-US" Id="csharp" Version="1.23.16" Publisher="ms-dotnettools"/>
```

And I Tried download from Marketplace manually.

![スクリーンショット 2021-12-06 12 04 42](https://user-images.githubusercontent.com/1955233/144780566-959be276-94ec-4994-b63b-550a93895750.png)

1. unzip ms-dotnettools.csharp-1.23.17@darwin-x64.vsix
2. grep 1.23.16 extension.vsixmanifest
3. grep 1.23.17 extension.vsixmanifest

```
    <Identity Language="en-US" Id="csharp" Version="1.23.17" Publisher="ms-dotnettools"  TargetPlatform="darwin-x64"  />
```

Is this problem caused by forgetting to update the version of the common package?
Please fix and re-publish it.

**Comment by yamachu** [comment]
Workaround

1. Download platform specified extension from https://marketplace.visualstudio.com/items?itemName=ms-dotnettools.csharp
2. Click Install from VSIX... and select vsix file that downloaded in step 1

<img width="499" alt="スクリーンショット 2021-12-06 12 29 42" src="https://user-images.githubusercontent.com/1955233/144782678-59434950-bee8-4b79-9187-9438873af0d0.png">

**Comment by jtsom** [comment]
You may want to check the log and the version actually installed. As I posted above it doesn't work manually installing.

**Comment by yamachu** [comment]
```
Starting OmniSharp server at 2021/12/6 13:00:54
    Target: /Users/yk-yamada/Projects/github.com/yamachu/BlazorWithReactSample/BlazorWebAssembly

OmniSharp server started.
    Path: /Users/yk-yamada/.vscode-insiders/extensions/ms-dotnettools.csharp-1.23.17/.omnisharp/1.37.17/run
    PID: 63277

[info]: OmniSharp.Stdio.Host
        Starting OmniSharp on MacOS 10.15.7 (x64)
[info]: OmniSharp.Services.DotNetCliService
        Checking the 'DOTNET_ROOT' environment variable to find a .NET SDK
[info]: OmniSharp.Services.DotNetCliService
        Using the 'dotnet' on the PATH.
[info]: OmniSharp.Services.DotNetCliService
        DotNetPath set to dotnet
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Located 1 MSBuild instance(s)
            1: StandAlone 17.0.0 - "/Users/yk-yamada/.vscode-insiders/extensions/ms-dotnettools.csharp-1.23.17/.omnisharp/1.37.17/omnisharp/.msbuild/Current/Bin"
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        MSBUILD_EXE_PATH environment variable set to '/Users/yk-yamada/.vscode-insiders/extensions/ms-dotnettools.csharp-1.23.17/.omnisharp/1.37.17/omnisharp/.msbuild/Current/Bin/MSBuild.exe'
[info]: OmniSharp.MSBuild.Discovery.MSBuildLocator
        Registered MSBuild instance: StandAlone 17.0.0 - "/Users/yk-yamada/.vscode-insiders/extensions/ms-dotnettools.csharp-1.23.17/.omnisharp/1.37.17/omnisharp/.msbuild/Current/Bin"
            CscToolExe = csc.exe
            MSBuildToolsPath = /Users/yk-yamada/.vscode-insiders/extensions/ms-dotnettools.csharp-1.23.17/.omnisharp/1.37.17/omnisharp/.msbuild/Current/Bin
            CscToolPath = /Users/yk-yamada/.vscode-insiders/extensions/ms-dotnettools.csharp-1.23.17/.omnisharp/1.37.17/omnisharp/.msbuild/Current/Bin/Roslyn
            BypassFrameworkInstallChecks = true
            MSBuildExtensionsPath = /Users/yk-yamada/.vscode-insiders/extensions/ms-dotnettools.csharp-1.23.17/.omnisharp/1.37.17/omnisharp/.msbuild
```

Works correctly.
I tried Windows 10 VSCode Insiders 1.63.0 d09289a2b... , it works with this workaround.
https://github.com/OmniSharp/omnisharp-vscode/issues/4922#issuecomment-986406186

**Comment by isidorn** [comment]
Hi, VS Code team member here 👋 
Sorry about this breakage and thank you very much for the investigation and for using VS Code Insiders 👏 
@sandy081, @saikanth007 and me investigated and the root cause for this is VS Code changing `header` values when sending requests to the MP. For this release, we have decided to revert back to using the old `header` values. This issue captures that and it will be available in the next VS Code insiders later today https://github.com/microsoft/vscode/issues/138503

For a deeper investigation why the change of the `user-agent` header influenced the difference `cdn` link we have created this issue on the Marketplace side https://github.com/microsoft/vsmarketplace/issues/267
Since the real root causes was wrong cdn link to an extension `manifest` file that pointed to an outdated extension version.
